homeassistant/components/devolo_home_control/config_flow.py

```diff
@@ -52,18 +53,19 @@ async def async_step_zeroconf(self, discovery_info: DiscoveryInfoType):
 
     async def async_step_zeroconf_confirm(self, user_input=None):
         """Handle a flow initiated by zeroconf."""
+        step_id = "zeroconf_confirm"
         if user_input is None:
-            return self._show_form(step_id="zeroconf_confirm")
-        return await self._connect_mydevolo(user_input)
+            return self._show_form(step_id=step_id)
+        return await self._connect_mydevolo(user_input, step_id)
 
-    async def _connect_mydevolo(self, user_input):
+    async def _connect_mydevolo(self, user_input, step_id):
         """Connect to mydevolo."""
         mydevolo = configure_mydevolo(conf=user_input)
         credentials_valid = await self.hass.async_add_executor_job(
             mydevolo.credentials_valid
         )
         if not credentials_valid:
-            return self._show_form({"base": "invalid_auth"})
+            return self._show_form(step_id=step_id, errors={"base": "invalid_auth"})
```

Comment:
It would be better to raise a custom exception here that we catch in the step method and show the form from the step method. We then don't need to pass arguments around as much and can keep the form scope contained in the respective step. Config flow helper methods should generally not show forms.

Reply:
Good point. I also needed to extend the tests for that to keep the coverage up.